server/src/main/java/org/elasticsearch/index/mapper/KeyedJsonAtomicFieldData.java

```diff
@@ -0,0 +1,237 @@
+        BytesRef extractedKey = JsonFieldParser.extractKey(term);
+        return key.compareTo(extractedKey);
+    }
+
+    private static class KeyedJsonDocValues extends AbstractSortedSetDocValues {
+
+        private final BytesRef key;
+        private final SortedSetDocValues delegate;
+
+        /**
+         * The first and last ordinals whose term has 'key' as a prefix. These
+         * values must be non-negative (there is at least one matching term).
+         */
+        private final long minOrd;
+        private final long maxOrd;
+
+        /**
+         * We cache the first ordinal in a document to avoid unnecessary iterations
+         * through the delegate doc values. If no ordinal is cached for the current
+         * document, this value will be -1.
+         */
+        private long cachedNextOrd;
+
+        private KeyedJsonDocValues(BytesRef key,
+                                   SortedSetDocValues delegate,
+                                   long minOrd,
+                                   long maxOrd) {
+            this.key = key;
+            this.delegate = delegate;
+            this.minOrd = minOrd;
+            this.maxOrd = maxOrd;
+            this.cachedNextOrd = -1;
+        }
+
+        @Override
+        public long getValueCount() {
+            return delegate.getValueCount();
+        }
+
+        /**
+         * Returns the (un-prefixed) term value for the requested ordinal.
+         *
+         * Note that this method can only be called on ordinals returned from {@link #nextOrd()}.
+         */
+        @Override
+        public BytesRef lookupOrd(long ord) throws IOException {
+            if (ord < minOrd || ord > maxOrd) {
+                throw new IllegalArgumentException("The provided ordinal [" + ord + "] is outside the valid " +
+                    "range. For keyed JSON fields, only ordinals returned from nextOrd can be passed to lookupOrd.");
+            }
+            BytesRef keyedValue = delegate.lookupOrd(ord);
+            int prefixLength = key.length + 1;
+            int valueLength = keyedValue.length - prefixLength;
+            return new BytesRef(keyedValue.bytes, prefixLength, valueLength);
+        }
+
+        @Override
+        public long nextOrd() throws IOException {
+            if (cachedNextOrd >= 0) {
+                long nextOrd = cachedNextOrd;
+                cachedNextOrd = -1;
+                return nextOrd;
+            }
+
+            long ord = delegate.nextOrd();
+            if (ord != NO_MORE_ORDS && ord <= maxOrd) {
+                return ord;
+            } else {
+                return NO_MORE_ORDS;
+            }
+        }
+
+        @Override
+        public boolean advanceExact(int target) throws IOException {
+            if (delegate.advanceExact(target)) {
+                for (long ord = delegate.nextOrd(); ord != NO_MORE_ORDS; ord = delegate.nextOrd()) {
+                     if (minOrd <= ord && ord <= maxOrd) {
```

Comment:
nit: we could actually break the loop if ord > maxOrd

Reply:
👍